User Defined DUID

marjohn56 opened the pull request:
User may define a DUID to use in System->Advanced->Networking. The
entered DUID is validated for composition and length, if valid it is
stored in the config.xml. On call of wan_dhcp6_configure() the DUID is
written to file to be read by dhcp6c on launch.

marjohn56 commented:
Simples... Oh that's a UK thing you may not understand! :)

jim-p commented on src/usr/local/www/system_advanced_network.php:
```diff
@@ -106,8 +119,8 @@
 		} else {
-			unset($config['system']['disablelargereceiveoffloading']);
-		}
+			unset($config['system']['disablelargereceiveoffloading']); hi		
```
What's this?

marjohn56 replied:
Not in my commit jim-p, don't know where that has come from.

jim-p replied:
It was introduced by the second commit on this PR, aff01dbd97c10731b98bc70c32ac19822c95d304

jim-p commented on src/usr/local/www/system_advanced_network.php:
```diff
@@ -186,6 +199,15 @@
+	$pconfig['global-v6duid'],
+	['placeholder' => 'xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx']
+	))->setWidth(9)->sethelp('Enter the DUID to use here. Use this option if using RAM Disk, as the DUID will be lost on reboot. The existing DUID may be found in var/db/dhcp6_duid.' .'<br />' .
+			'NOTE: Do not use this option with multiple DHCP6 WAN interfaces.');
```
Is this note still relevant? If it's a system-wide DUID it should be safe to use for any number of interfaces.

marjohn56 replied:
I only have one WAN interface so I cannot be sure about this, and I know what we said about DUID the other day, but the RFC does say that DUID's should be interface unique. When running dhcp6c on multiple WAN interfaces I assume multiple dhcp6c clients are launched, or is that not the case?

If it is the case then the solution is to go back to the by WAN setting of DUID's and I'll modify dhcp6c to use specific DUIDs for each interface. We already pass an the argument to dhcp6c anyway so it would be very simple to do.

jim-p replied:
The RFC doesn't say that. None that I've seen anyhow. While the DUID can be generated using an interface address as part of the value, DUID is not intended to be related to any interface. It's a client identifier that uniquely identifies a _device_ and not an interface. Identifying specific interfaces is where IAIDs come in, which we've been over.

marjohn56 replied:
I will bow to your greater knowledge on that then.

Problem is, on RAM disks the DUID disappears on reboot, a new one is created, but that only happens when dhcp6 is fired up, I can see all sorts of 'I've got no DUID' type questions.  I'll have a look at pulling the duid from the file and inserting it in the placeholder. If nothing else I'm finding out more about pfSense..

jim-p replied:
That easy, if there is no file, just print the x's.

marjohn56 replied:
True..

marjohn56 replied:
Completed.. Existing DUID now displays in placeholder. I have put more verbage there to try and let users know that unless they enter the duid and save it, it is not stored in the config and therefore can change, especially when using RAM disks.

jim-p commented on src/usr/local/www/system_advanced_network.php:
```diff
@@ -186,6 +199,15 @@
+	'text',
+	$pconfig['global-v6duid'],
+	['placeholder' => 'xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx']
+	))->setWidth(9)->sethelp('Enter the DUID to use here. Use this option if using RAM Disk, as the DUID will be lost on reboot. The existing DUID may be found in var/db/dhcp6_duid.' .'<br />' .
```
Might be nice to print and/or set the placeholder to the current DUID rather than suggest the user dig into a system file to find it.

marjohn56 replied:
O.K. If the system is virginal, there will not be any existing DUID, but that's obvious. If we set the placeholder, do we need to add a checkbox to say use the DUID? Not sure how the placeholder works, do you need to physically enter what is there as the placeholder in order for it to take?

jim-p replied:
If they leave the field blank it will use the existing DUID, which is what the placeholder is at the time. There wouldn't be a way to make it editable from there but usually it wouldn't need to be. If someone is changing it they likely would be changing it to a much different value.

marjohn56 replied:
But what happens if they don't want to use the Global DUID?

jim-p replied:
If they want to change it, then they enter in whatever they want. It's a convenience there to show the user what the system is using now rather than making them dig for it. The placeholder value is displayed in the field grayed out to indicate the "default" value of the field.

marjohn56 commented:
Looks like I was tying an email at the same time... :(

 

Corrected and updated.

 

From: Jim P [mailto:notifications@github.com] 
Sent: 22 December 2016 13:42
To: pfsense/pfsense <pfsense@noreply.github.com>
Cc: marjohn56 <martin@queens-park.com>; Author <author@noreply.github.com>
Subject: Re: [pfsense/pfsense] User Defined DUID (#3304)

 

@jim-p commented on this pull request.

  _____  

In src/usr/local/www/system_advanced_network.php <https://github.com/pfsense/pfsense/pull/3304> :

> @@ -106,8 +119,8 @@
               if ($_POST['disablelargereceiveoffloading'] == "yes") {
                       $config['system']['disablelargereceiveoffloading'] = true;
               } else {
-                      unset($config['system']['disablelargereceiveoffloading']);
-              }
+                      unset($config['system']['disablelargereceiveoffloading']); hi             

It was introduced by the second commit on this PR,  <https://github.com/pfsense/pfsense/commit/aff01dbd97c10731b98bc70c32ac19822c95d304> aff01db

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/pfsense/pfsense/pull/3304> , or mute the thread <https://github.com/notifications/unsubscribe-auth/ARIXphlI4d-GNvFEox9XlYvRErWVm6a4ks5rKn4fgaJpZM4LTEDL> .  <https://github.com/notifications/beacon/ARIXppE4cEWTnRtZmh0BZZuFOyqn4mzBks5rKn4fgaJpZM4LTEDL.gif>

marjohn56 commented:
True.. :)

rbgarga commented on src/etc/inc/util.inc:
```diff
@@ -2542,4 +2542,67 @@ function validateipaddr(&$addr, $type, $label, &$err_msg, $alias=false) {
+	log_error(gettext("Error: attempting to write DUID file - File write error"));
+	return false;
+}
+/* returns duid string from 'vardb_path']}/dhcp6c_duid' */
```
Add a blank line between functions

rbgarga commented on src/etc/inc/util.inc:
```diff
@@ -2542,4 +2542,67 @@ function validateipaddr(&$addr, $type, $label, &$err_msg, $alias=false) {
+	
+	if ($fd = fopen("{$g['vardb_path']}/dhcp6c_duid", "r")) {
+		if(filesize("{$g['vardb_path']}/dhcp6c_duid")==16) {
+			$buffer = fread($fd,filesize("{$g['vardb_path']}/dhcp6c_duid"));					
```
At this point you are sure filesize() returned 16 so you don't need to call it again. Instead use 16

rbgarga commented on src/etc/inc/util.inc:
```diff
@@ -2542,4 +2542,67 @@ function validateipaddr(&$addr, $type, $label, &$err_msg, $alias=false) {
+{
+	global $g;
+	
+	$count = 0;
```
Initialize $duid_ASCII

rbgarga commented on src/etc/inc/util.inc:
```diff
@@ -2542,4 +2542,67 @@ function validateipaddr(&$addr, $type, $label, &$err_msg, $alias=false) {
+			$buffer = fread($fd,filesize("{$g['vardb_path']}/dhcp6c_duid"));					
+			while($count < 16) {				
+				$a = $buffer[$count];
+				$duid_ASCII .= bin2hex($a);
```
There is no need to set this variable $a, you can pass $buffer[$count] as bin2hex() parameter. And also, $a is not a meaningful variable name

rbgarga commented on src/etc/inc/util.inc:
```diff
@@ -2542,4 +2542,67 @@ function validateipaddr(&$addr, $type, $label, &$err_msg, $alias=false) {
+		fclose($fd);
+	}
+	//if no file or error with read then the string returns blanked DUID string
+	if($count != 16) {
```
Instead of check for $count it would be better to check using is_duid() to be sure it's a correct value

rbgarga commented on src/usr/local/www/system_advanced_network.php:
```diff
@@ -83,6 +84,18 @@
+				$config['system']['global-v6duid'] = $_POST['global-v6duid'];
+			}
+		}
+		else {
```
else goes in the same line of } as in:
```
} else {
```

marjohn56 replied:
Done

phil-davis commented on src/usr/local/www/system_advanced_network.php:
```diff
@@ -186,6 +199,17 @@
+	'text',
+	$pconfig['global-v6duid'],
+	['placeholder' => $duid]
+	))->setWidth(9)->sethelp('The current DUID is displayed above. You may enter a new DUID whuch will be used on the next WAN interface UP event.' .'<br />' .
```
s/whuch/which

marjohn56 commented:
Hurrah! :)

 

From: Renato Botelho [mailto:notifications@github.com] 
Sent: 29 December 2016 12:44
To: pfsense/pfsense <pfsense@noreply.github.com>
Cc: marjohn56 <martin@queens-park.com>; Author <author@noreply.github.com>
Subject: Re: [pfsense/pfsense] User Defined DUID (#3304)

 

@rbgarga approved this pull request.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/pfsense/pfsense/pull/3304#pullrequestreview-14662804> , or mute the thread <https://github.com/notifications/unsubscribe-auth/ARIXph9rrjQzXbmawKfLYLPpYFjy0qsvks5rM6sFgaJpZM4LTEDL> .  <https://github.com/notifications/beacon/ARIXpiVfB8DUwkxYmz3RupUlGiH7FAeoks5rM6sFgaJpZM4LTEDL.gif>

chrcoluk commented:
I noticed an issue.

The code enforces a 16 character length DUID.

However if you start dhcp6 client in debug mode (so it prints current DUID to log), the DUID pfsense generates is only 14 characters long.  So lets say you want a persistent DUID and you paste in the most recent generated pfsense DUID, the box will report the DUID is invalid when its not, as DUID does not have a minimum length of 16 in the specs.

I did notice the ghosted out example DUID started of with 4 extra characters of 0e:00 which I entered and it worked, but those 4 characters are not part of the dynamic DUID generation.

Also its more logical to place this box in the WAN IPv6 section not generic networking as it only applies on wan_dhcp6.

phil-davis commented:
A DUID can be generated in a number of ways, and is variable length up to 128 octets:
https://tools.ietf.org/html/rfc3315#page-19
https://tools.ietf.org/html/rfc6355

marjohn56 commented:
@chrcoluk It;s not a ghosted out example., The DUID is read in from the existing DUID generated by dhcp6c, If you open the DUID in a hex editor you would see that. And we had a very long discussion as to where the DUID should be. As it's a GLOBAL option, i.e.  it applies system wide then it lives where it lives, otherwise you could have multiple WAN's with difference DUID's, and that''s incorrect.

@phil-davis Spot on!

marjohn56 commented:
just had a look at dhcp6c, it only accepts and generates a type 1 DUID - Link-layer address plus time (DUID-LLT).

phil-davis commented:
See #3365 for fixes to where the input is validated, and a format_duid() helper function that helps massage various possible user input into the required format.
